lib/Dialect/FIRRTL/InnerSymbolTable.cpp

```diff
@@ -26,54 +26,87 @@ namespace firrtl {
 //===----------------------------------------------------------------------===//
 // InnerSymbolTable
 //===----------------------------------------------------------------------===//
-
 InnerSymbolTable::InnerSymbolTable(Operation *op) {
-  using llvm::dbgs;
-  LLVM_DEBUG(dbgs() << "===----- InnerSymbolTable -----===\n";
-             dbgs() << "Constructing table for @"
-                    << SymbolTable::getSymbolName(op) << "\n");
-  assert(op->hasTrait<OpTrait::InnerSymbolTable>() &&
-         "expected operation to have InnerSymbolTable trait");
+  assert(op->hasTrait<OpTrait::InnerSymbolTable>());
   // Save the operation this table is for.
   this->innerSymTblOp = op;
 
+  (void)walkSymbols(op, [&](StringAttr name, InnerSymTarget target) {
+    auto it = symbolTable.insert({name, target});
+    (void)it;
+    assert(it.second && "repeated symbol found");
+    return success();
+  });
+}
+
+FailureOr<InnerSymbolTable> InnerSymbolTable::get(Operation *op) {
+  assert(op);
+  if (!op->hasTrait<OpTrait::InnerSymbolTable>())
+    return op->emitError("expected operation to have InnerSymbolTable trait");
+
+  TableTy table;
+  auto result = walkSymbols(
+      op, [&](StringAttr name, InnerSymTarget target) -> LogicalResult {
+        auto it = table.insert({name, target});
+        if (it.second)
+          return success();
+        auto existing = it.first->second;
+        return target.getOp()
+            ->emitError()
+            .append("redefinition of inner symbol named '", name.strref(), "'")
+            .attachNote(existing.getOp()->getLoc())
+            .append("see existing inner symbol definition here");
+      });
+  if (failed(result))
+    return failure();
+  return InnerSymbolTable(op, std::move(table));
+};
+
+LogicalResult InnerSymbolTable::walkSymbols(Operation *op,
+                                            InnerSymCallbackFn callback) {
+  using llvm::dbgs;
+  LLVM_DEBUG(dbgs() << "===----- InnerSymbolTable -----===\n";
+             dbgs() << "Walking inner symbols for @"
+                    << SymbolTable::getSymbolName(op) << "\n");
   auto addSym = [&](StringAttr name, InnerSymTarget target) {
     LLVM_DEBUG(dbgs() << " - @" << name << " -> " << target << "\n");
     assert(name && !name.getValue().empty());
```

Comment:
Assert liberally! But yes it would be better for them to be correct by construction.  This was moreso to catch my own mistake than accommodate invalid Attr's.  Although presently we have empty strings for ports that don't have symbols (if at least one port has a valid symbol, else the array is empty).

Maybe more reason to push this into InnerSymbolOpInterface / IST methods that callback with valid symbols/targets, to keep those details isolated there!